File: apps/api-server/src/modules/shared_kernel/error_envelope.py
from __future__ import annotations

from typing import Any


class APIError(Exception):
    def __init__(
        self,
        code: str,
        message: str,
        details: Any = None,
        status_code: int = 400,
    ) -> None:
        self.code = code
        self.message = message
        self.details = details
        self.status_code = status_code
        super().__init__(message)


class ErrorEnvelope:
    @staticmethod
    def from_error(err: APIError) -> dict[str, dict[str, Any]]:
        return {
            "error": {
                "code": err.code,
                "message": err.message,
                "details": err.details,
            }
        }

File: apps/web-dashboard/src/shared/api-client/errorEnvelope.ts
export interface ApiError {
  code: string;
  message: string;
  details?: unknown;
}

export function isApiError(body: unknown): body is { error: ApiError } {
  if (typeof body !== "object" || body === null || !("error" in body)) {
    return false;
  }

  const error = (body as Record<string, unknown>).error;
  return (
    typeof error === "object" &&
    error !== null &&
    typeof (error as Record<string, unknown>).code === "string" &&
    typeof (error as Record<string, unknown>).message === "string"
  );
}

export function formatErrorMessage(error: unknown): string {
  if (error instanceof Error) return error.message;
  if (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  ) {
    return (error as { message: string }).message;
  }
  return "Unknown error";
}
